text/003-trigger-authority.md

```diff
@@ -0,0 +1,134 @@
+  they can have administrative permissions, react on some state changes made by an untrusted source, and execute
+  administrative actions.
+- **Use Trigger Completed filter**. Then, an un-trusted source could call some un-trusted trigger that would create some
+  state change, and then the administrative trigger would execute as a result of that trigger completion, performing
+  actions with elevated permissions. This works essentially the same as the Pipeline/Time variant, but with immediate
+  execution.
+- **Use executor-level custom instructions**. Executor has "root privileges" and can perform any action. However, it is
+  not a modular and composable approach, as there is only one global executor in Iroha. With triggers, it is possible
+  for different unrelated authorities to deploy their own permission-ed logic without interfering with each other. It's
+  just more modular, come on!
+
+### Use case 2
+
+I (e.g. Alice) want to deploy a data trigger that reacts on someone else's (e.g. Bob's) asset balance change. This
+trigger would, for example, make a transfer on _my_ behalf.
+
+However, when Bob changes his own balance, my trigger would be executed with _Bob's_ permissions, not mine, and
+therefore the trigger will fail to make the intended transfer.
+
+## Proposal
+
+Current model--always run with caller's authority--is:
+
+- Maximally safe - a trigger never does more than the caller can
+- But useless for efficient privileged automations
+- Reduces by-call and data triggers to extensions of the caller's transactions without autonomy
+
+I see two options.
+
+### Option 1 - Always run with trigger authority
+
+- Simple mental model: trigger is an autonomous actor.
+- Security: developers must explicitly check caller identity inside the trigger logic. This isn't possible in case of
+  simple `Executable::Instructions`.
+
+This is close to how Ethereum smart contracts work, AFAIK.
+
+### Option 2 - Mixed model
+
+- Provide an ability to set whether a trigger should use a **pre-defined authority** or the **caller's** authority.
+- This would apply to Data, By-Call, and Trigger-Completed triggers. Time and Pipeline triggers are "called" by the
+  host, so they must always have a pre-defined authority.
+
+Definition example:
+
+```rust
+enum VariableAuthority {
+  /// Always use this account
+  Specified(AccountId),
+  /// Use the caller's authority
+  Caller,
+}
+```
+
+## Consideration: triggers are a part of transaction?
+
+Currently, data triggers are a part of the original transaction, and if data trigger fails - the entire transaction
+fails.
+
+This seems to be useful in some cases, but is not compatible with the case when Data trigger has a pre-defined
+authority: Alice then could fail Bob's transactions if she wants. However, it may also be desirable: an admin may want
+to fail someone's transaction by their data trigger.
```

Comment:
The control of whether someone’s transaction succeeds or fails should be centralized in the permission system.

Reply:
Why?

Permission system allows for fine-grained access to resources and actions. I am not sure I understand why it _should_ be centralized.

Reply:
Historically, the permission system only governed state read/write access.
Once we bring attach/detach into scope, the permission model must also cover a second dimension: transaction scope control (propagation reach and generational boundaries):

- Dimension 1: State Access (read/write capabilities)
- Dimension 2: Scope Control (changing the propagation surface via attach/detach)